File: output/GPT3.5-Turbo/xCodeEval_4000/Python/Go/5cb6c1b3fc0e089985a249812459323b.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

var mod int = 1000000007
var mod1 int = 998244353

type TreeNode struct {
	key       int
	value     int
	left      *TreeNode
	right     *TreeNode
	parent    *TreeNode
	height    int
	num_left  int
	num_total int
}

type AvlTree struct {
	tree *TreeNode
}

func (t *AvlTree) add(k, v int) {
	if t.tree == nil {
		t.tree = &TreeNode{
			key:       k,
			value:     v,
			left:      nil,
			right:     nil,
			parent:    nil,
			height:    1,
			num_left:  1,
			num_total: 1,
		}
		return
	}
	node := t._add(k, v)
	if node != nil {
		t._rebalance(node)
	}
}

func (t *AvlTree) _add(k, v int) *TreeNode {
	node := t.tree
	for node != nil {
		if k < node.key {
			if node.left != nil {
				node = node.left
			} else {
				node.left = &TreeNode{
					key:       k,
					value:     v,
					left:      nil,
					right:     nil,
					parent:    node,
					height:    1,
					num_left:  1,
					num_total: 1,
				}
				return node.left
			}
		} else if node.key < k {
			if node.right != nil {
				node = node.right
			} else {
				node.right = &TreeNode{
					key:       k,
					value:     v,
					left:      nil,
					right:     nil,
					parent:    node,
					height:    1,
					num_left:  1,
					num_total: 1,
				}
				return node.right
			}
		} else {
			node.value = v
			return nil
		}
	}
	return nil
}

func (t *AvlTree) get_height(x *TreeNode) int {
	if x != nil {
		return x.height
	}
	return 0
}

func (t *AvlTree) get_num_total(x *TreeNode) int {
	if x != nil {
		return x.num_total
	}
	return 0
}

func (t *AvlTree) _rebalance(node *TreeNode) {
	n := node
	for n != nil {
		lh := t.get_height(n.left)
		rh := t.get_height(n.right)
		n.height = max(lh, rh) + 1
		balance_factor := lh - rh
		n.num_total = 1 + t.get_num_total(n.left) + t.get_num_total(n.right)
		n.num_left = 1 + t.get_num_total(n.left)

		if balance_factor > 1 {
			if t.get_height(n.left.left) < t.get_height(n.left.right) {
				t._rotate_left(n.left)
			}
			t._rotate_right(n)
		} else if balance_factor < -1 {
			if t.get_height(n.right.right) < t.get_height(n.right.left) {
				t._rotate_right(n.right)
			}
			t._rotate_left(n)
		} else {
			n = n.parent
		}
	}
}

func (t *AvlTree) _remove_one(node *TreeNode) {
	replacement := node.left
	if replacement == nil {
		replacement = node.right
	}
	if node.parent != nil {
		if t._is_left(node) {
			node.parent.left = replacement
		} else {
			node.parent.right = replacement
		}
		if replacement != nil {
			replacement.parent = node.parent
		}
		node.parent = nil
	} else {
		t.tree = replacement
		if replacement != nil {
			replacement.parent = nil
		}
	}
	node.left = nil
	node.right = nil
	node.parent = nil
	t._rebalance(replacement)
}

func (t *AvlTree) _remove_leaf(node *TreeNode) {
	if node.parent != nil {
		if t._is_left(node) {
			node.parent.left = nil
		} else {
			node.parent.right = nil
		}
		t._rebalance(node.parent)
	} else {
		t.tree = nil
	}
	node.parent = nil
	node.left = nil
	node.right = nil
}

func (t *AvlTree) remove(k int) {
	node := t._get_node(k)
	if node == nil {
		return
	}
	if t._is_leaf(node) {
		t._remove_leaf(node)
		return
	}
	if node.left != nil && node.right != nil {
		nxt := t._get_next(node)
		node.key = nxt.key
		node.value = nxt.value
		if t._is_leaf(nxt) {
			t._remove_leaf(nxt)
		} else {
			t._remove_one(nxt)
		}
		t._rebalance(node)
	} else {
		t._remove_one(node)
	}
}

func (t *AvlTree) get(k int) int {
	node := t._get_node(k)
	if node != nil {
		return node.value
	}
	return -1
}

func (t *AvlTree) _get_node(k int) *TreeNode {
	node := t.tree
	for node != nil {
		if k < node.key {
			node = node.left
		} else if node.key < k {
			node = node.right
		} else {
			return node
		}
	}
	return nil
}

func (t *AvlTree) get_at(pos int) (int, int) {
	x := pos + 1
	node := t.tree
	for node != nil {
		if x < node.num_left {
			node = node.left
		} else if node.num_left < x {
			x -= node.num_left
			node = node.right
		} else {
			return node.key, node.value
		}
	}
	panic("Out of ranges")
}

func (t *AvlTree) _is_left(node *TreeNode) bool {
	return node.parent != nil && node.parent.left == node
}

func (t *AvlTree) _is_leaf(node *TreeNode) bool {
	return node.left == nil && node.right == nil
}

func (t *AvlTree) _rotate_right(node *TreeNode) {
	if node.parent == nil {
		t.tree = node.left
		node.left.parent = nil
	} else if t._is_left(node) {
		node.parent.left = node.left
		node.left.parent = node.parent
	} else {
		node.parent.right = node.left
		node.left.parent = node.parent
	}
	bk := node.left.right
	node.left.right = node
	node.parent = node.left
	node.left = bk
	if bk != nil {
		bk.parent = node
	}
	node.height = max(t.get_height(node.left), t.get_height(node.right)) + 1
	node.num_total = 1 + t.get_num_total(node.left) + t.get_num_total(node.right)
	node.num_left = 1 + t.get_num_total(node.left)
}

func (t *AvlTree) _rotate_left(node *TreeNode) {
	if node.parent == nil {
		t.tree = node.right
		node.right.parent = nil
	} else if t._is_left(node) {
		node.parent.left = node.right
		node.right.parent = node.parent
	} else {
		node.parent.right = node.right
		node.right.parent = node.parent
	}
	bk := node.right.left
	node.right.left = node
	node.parent = node.right
	node.right = bk
	if bk != nil {
		bk.parent = node
	}
	node.height = max(t.get_height(node.left), t.get_height(node.right)) + 1
	node.num_total = 1 + t.get_num_total(node.left) + t.get_num_total(node.right)
	node.num_left = 1 + t.get_num_total(node.left)
}

func (t *AvlTree) _get_next(node *TreeNode) *TreeNode {
	if node.right == nil {
		return node.parent
	}
	n := node.right
	for n.left != nil {
		n = n.left
	}
	return n
}

type SegmentTree1 struct {
	data    []int
	_size   int
	_default int
	_func    func(int, int) int
}

func NewSegmentTree1(data []int, defaultVal int, funcVal func(int, int) int) *SegmentTree1 {
	size := 1 << (len(data) - 1)
	tree := &SegmentTree1{
		data:    make([]int, 2*size),
		_size:   size,
		_default: defaultVal,
		_func:    funcVal,
	}
	copy(tree.data[size:size+len(data)], data)
	for i := size - 1; i >= 0; i-- {
		tree.data[i] = tree._func(tree.data[2*i], tree.data[2*i+1])
	}
	return tree
}

func (t *SegmentTree1) Query(start, stop int) int {
	if start == stop {
		return t.data[start+t._size]
	}
	stop++
	start += t._size
	stop += t._size

	res := t._default
	for start < stop {
		if start&1 == 1 {
			res = t._func(res, t.data[start])
			start++
		}
		if stop&1 == 1 {
			stop--
			res = t._func(res, t.data[stop])
		}
		start >>= 1
		stop >>= 1
	}
	return res
}

type SegmentTree struct {
	data    []int
	_size   int
	_default int
	_func    func(int, int) int
}

func NewSegmentTree(data []int, defaultVal int, funcVal func(int, int) int) *SegmentTree {
	size := 1 << (len(data) - 1)
	tree := &SegmentTree{
		data:    make([]int, 2*size),
		_size:   size,
		_default: defaultVal,
		_func:    funcVal,
	}
	copy(tree.data[size:size+len(data)], data)
	for i := size - 1; i >= 0; i-- {
		tree.data[i] = tree._func(tree.data[2*i], tree.data[2*i+1])
	}
	return tree
}

func (t *SegmentTree) Query(start, stop int) int {
	if start == stop {
		return t.data[start+t._size]
	}
	stop++
	start += t._size
	stop += t._size

	res := t._default
	for start < stop {
		if start&1 == 1 {
			res = t._func(res, t.data[start])
			start++
		}
		if stop&1 == 1 {
			stop--
			res = t._func(res, t.data[stop])
		}
		start >>= 1
		stop >>= 1
	}
	return res
}

type Factorial struct {
	MOD            int
	factorials     []int
	invModulos     []int
	invFactorial_  []int
}

func NewFactorial(MOD int) *Factorial {
	f := &Factorial{
		MOD: MOD,
		factorials: []int{1, 1},
		invModulos: []int{0, 1},
		invFactorial_: []int{1, 1},
	}
	return f
}

func (f *Factorial) Calc(n int) int {
	if n <= -1 {
		fmt.Println("Invalid argument to calculate n!")
		fmt.Println("n must be non-negative value. But the argument was", n)
		os.Exit(1)
	}
	if n < len(f.factorials) {
		return f.factorials[n]
	}
	nextArr := make([]int, n+1-len(f.factorials))
	initialI := len(f.factorials)
	prev := f.factorials[len(f.factorials)-1]
	m := f.MOD
	for i := initialI; i <= n; i++ {
		prev = nextArr[i-initialI] = (prev * i) % m
	}
	f.factorials = append(f.factorials, nextArr...)
	return f.factorials[n]
}

func (f *Factorial) Inv(n int) int {
	if n <= -1 {
		fmt.Println("Invalid argument to calculate n^(-1)")
		fmt.Println("n must be non-negative value. But the argument was", n)
		os.Exit(1)
	}
	p := f.MOD
	pi := n % p
	if pi < len(f.invModulos) {
		return f.invModulos[pi]
	}
	nextArr := make([]int, n+1-len(f.invModulos))
	initialI := len(f.invModulos)
	for i := initialI; i < p && i <= n; i++ {
		next := -f.invModulos[p%i] * (p / i) % p
		f.invModulos = append(f.invModulos, next)
	}
	return f.invModulos[pi]
}

func (f *Factorial) InvFactorial(n int) int {
	if n <= -1 {
		fmt.Println("Invalid argument to calculate (n^(-1))!")
		fmt.Println("n must be non-negative value. But the argument was", n)
		os.Exit(1)
	}
	if n < len(f.invFactorial_) {
		return f.invFactorial_[n]
	}
	f.Inv(n)
	nextArr := make([]int, n+1-len(f.invFactorial_))
	initialI := len(f.invFactorial_)
	prev := f.invFactorial_[len(f.invFactorial_)-1]
	p := f.MOD
	for i := initialI; i <= n; i++ {
		prev = nextArr[i-initialI] = (prev * f.invModulos[i%p]) % p
	}
	f.invFactorial_ = append(f.invFactorial_, nextArr...)
	return f.invFactorial_[n]
}

type Combination struct {
	MOD       int
	factorial *Factorial
}

func NewCombination(MOD int) *Combination {
	c := &Combination{
		MOD:       MOD,
		factorial: NewFactorial(MOD),
	}
	return c
}

func (c *Combination) Ncr(n, k int) int {
	if k < 0 || n < k {
		return 0
	}
	k = min(k, n-k)
	f := c.factorial
	return (f.Calc(n) * f.InvFactorial(max(n-k, k)) % c.MOD * f.InvFactorial(min(k, n-k)) % c.MOD) % c.MOD
}

func Powm(a, n, m int) int {
	if a == 1 || n == 0 {
		return 1
	}
	if n%2 == 0 {
		s := Powm(a, n/2, m)
		return (s * s) % m
	} else {
		return (a * Powm(a, n-1, m)) % m
	}
}

func SortList(list1, list2 []int) []int {
	zippedPairs := make([][]int, len(list2))
	for i := 0; i < len(list2); i++ {
		zippedPairs[i] = []int{list2[i], list1[i]}
	}
	sort.Slice(zippedPairs, func(i, j int) bool {
		return zippedPairs[i][0] < zippedPairs[j][0]
	})
	z := make([]int, len(list1))
	for i := 0; i < len(list1); i++ {
		z[i] = zippedPairs[i][1]
	}
	return z
}

func Product(l []int) int {
	por := 1
	for i := 0; i < len(l); i++ {
		por *= l[i]
	}
	return por
}

func BinarySearchCount(arr []int, n, key int) int {
	left := 0
	right := n - 1

	count := 0

	for left <= right {
		mid := (right + left) / 2

		if arr[mid] <= key {
			count = mid + 1
			left = mid + 1
		} else {
			right = mid - 1
		}
	}

	return count
}

func CountDig(n int) int {
	c := 0
	for n > 0 {
		n /= 10
		c++
	}
	return c
}

func CountGreater(arr []int, n, k int) int {
	l := 0
	r := n - 1

	leftGreater := n

	for l <= r {
		m := (l + r) / 2
		if arr[m] >= k {
			leftGreater = m
			r = m - 1
		} else {
			l = m + 1
		}
	}

	return (n - leftGreater)
}

func main() {
	reader := bufio.NewReader(os.Stdin)
	writer := bufio.NewWriter(os.Stdout)
	defer writer.Flush()

	var n, k int
	fmt.Fscan(reader, &n, &k)
	l := make([]int, n)
	for i := 0; i < n; i++ {
		fmt.Fscan(reader, &l[i])
		l[i] %= k
	}
	if n == 1 {
		fmt.Fprintln(writer, l[0])
		return
	}
	l1 := l[:n/2]
	l2 := l[n/2:]
	s := make(map[int]bool)
	s1 := make(map[int]bool)
	s2 := make(map[int]bool)
	for i := 0; i < n/2; i++ {
		for j := range s {
			s2[(j+l1[i])%k] = true
		}
		s[l1[i]] = true
		for j := range s2 {
			s[j] = true
		}
		s2 = make(map[int]bool)
	}
	for i := 0; i < n-n/2; i++ {
		for j := range s1 {
			s2[(j+l2[i])%k] = true
		}
		s1[l2[i]] = true
		for j := range s2 {
			s1[j] = true
		}
		s2 = make(map[int]bool)
	}
	sliceS := make([]int, 0)
	for key := range s {
		sliceS = append(sliceS, key)
	}
	sliceS1 := make([]int, 0)
	for key := range s1 {
		sliceS1 = append(sliceS1, key)
	}
	sort.Ints(sliceS)
	sort.Ints(sliceS1)
	ans := max(max(sliceS...), max(sliceS1...))
	for i := 0; i < len(sliceS); i++ {
		t := k - 1 - sliceS[i]
		ind := BinarySearchCount(sliceS1, len(sliceS1), t)
		r := sliceS1[ind-1]
		rt := sliceS1[len(sliceS1)-1]
		t = ans
		ans = max((r+sliceS[i])%k, (sliceS[i]+rt)%k)
		ans = max(ans, t)
	}

	fmt.Fprintln(writer, ans)
}

func max(nums ...int) int {
	maxNum := nums[0]
	for _, num := range nums {
		if num > maxNum {
			maxNum = num
		}
	}
	return maxNum
}

func min(nums ...int) int {
	minNum := nums[0]
	for _, num := range nums {
		if num < minNum {
			minNum = num
		}
	}
	return minNum
}

File: output/GPT4o-Mini/xCodeEval_4000/Python/Java/5cb6c1b3fc0e089985a249812459323b.java
import java.io.*;
import java.util.*;

class FastIO {
    private final InputStream inputStream;
    private final OutputStream outputStream;
    private final byte[] buffer = new byte[8192];
    private int bufferPointer, bytesRead;

    public FastIO(InputStream inputStream, OutputStream outputStream) {
        this.inputStream = inputStream;
        this.outputStream = outputStream;
    }

    public int read() {
        if (bufferPointer == bytesRead) {
            bufferPointer = 0;
            try {
                bytesRead = inputStream.read(buffer);
            } catch (IOException e) {
                return -1;
            }
            if (bytesRead == -1) return -1;
        }
        return buffer[bufferPointer++];
    }

    public String next() {
        StringBuilder sb = new StringBuilder();
        int c;
        while ((c = read()) != -1 && !Character.isWhitespace(c)) {
            sb.append((char) c);
        }
        return sb.toString();
    }

    public void print(String s) {
        try {
            outputStream.write(s.getBytes());
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void println(String s) {
        print(s + "\n");
    }

    public void close() throws IOException {
        inputStream.close();
        outputStream.close();
    }
}

class TreeNode {
    int key, value;
    TreeNode left, right, parent;
    int height, numLeft, numTotal;

    TreeNode(int k, int v) {
        this.key = k;
        this.value = v;
        this.height = 1;
        this.numLeft = 1;
        this.numTotal = 1;
    }
}

class AvlTree {
    private TreeNode tree;

    public void add(int k, int v) {
        if (tree == null) {
            tree = new TreeNode(k, v);
            return;
        }
        TreeNode node = _add(k, v);
        if (node != null) {
            _rebalance(node);
        }
    }

    private TreeNode _add(int k, int v) {
        TreeNode node = tree;
        while (node != null) {
            if (k < node.key) {
                if (node.left != null) {
                    node = node.left;
                } else {
                    node.left = new TreeNode(k, v);
                    node.left.parent = node;
                    return node.left;
                }
            } else if (node.key < k) {
                if (node.right != null) {
                    node = node.right;
                } else {
                    node.right = new TreeNode(k, v);
                    node.right.parent = node;
                    return node.right;
                }
            } else {
                node.value = v;
                return null;
            }
        }
        return null;
    }

    private static int getHeight(TreeNode x) {
        return x != null ? x.height : 0;
    }

    private static int getNumTotal(TreeNode x) {
        return x != null ? x.numTotal : 0;
    }

    private void _rebalance(TreeNode node) {
        TreeNode n = node;
        while (n != null) {
            int lh = getHeight(n.left);
            int rh = getHeight(n.right);
            n.height = Math.max(lh, rh) + 1;
            n.numTotal = 1 + getNumTotal(n.left) + getNumTotal(n.right);
            n.numLeft = 1 + getNumTotal(n.left);

            int balanceFactor = lh - rh;

            if (balanceFactor > 1) {
                if (getHeight(n.left.left) < getHeight(n.left.right)) {
                    _rotateLeft(n.left);
                }
                _rotateRight(n);
            } else if (balanceFactor < -1) {
                if (getHeight(n.right.right) < getHeight(n.right.left)) {
                    _rotateRight(n.right);
                }
                _rotateLeft(n);
            } else {
                n = n.parent;
            }
        }
    }

    private void _rotateRight(TreeNode node) {
        if (node.parent == null) {
            tree = node.left;
            node.left.parent = null;
        } else if (isLeft(node)) {
            node.parent.left = node.left;
            node.left.parent = node.parent;
        } else {
            node.parent.right = node.left;
            node.left.parent = node.parent;
        }
        TreeNode bk = node.left.right;
        node.left.right = node;
        node.parent = node.left;
        node.left = bk;
        if (bk != null) {
            bk.parent = node;
        }
        node.height = Math.max(getHeight(node.left), getHeight(node.right)) + 1;
        node.numTotal = 1 + getNumTotal(node.left) + getNumTotal(node.right);
        node.numLeft = 1 + getNumTotal(node.left);
    }

    private void _rotateLeft(TreeNode node) {
        if (node.parent == null) {
            tree = node.right;
            node.right.parent = null;
        } else if (isLeft(node)) {
            node.parent.left = node.right;
            node.right.parent = node.parent;
        } else {
            node.parent.right = node.right;
            node.right.parent = node.parent;
        }
        TreeNode bk = node.right.left;
        node.right.left = node;
        node.parent = node.right;
        node.right = bk;
        if (bk != null) {
            bk.parent = node;
        }
        node.height = Math.max(getHeight(node.left), getHeight(node.right)) + 1;
        node.numTotal = 1 + getNumTotal(node.left) + getNumTotal(node.right);
        node.numLeft = 1 + getNumTotal(node.left);
    }

    private static boolean isLeft(TreeNode node) {
        return node.parent.left == node;
    }

    public void remove(int k) {
        TreeNode node = _getNode(k);
        if (node == null) return;
        if (isLeaf(node)) {
            _removeLeaf(node);
            return;
        }
        if (node.left != null && node.right != null) {
            TreeNode nxt = _getNext(node);
            node.key = nxt.key;
            node.value = nxt.value;
            if (isLeaf(nxt)) {
                _removeLeaf(nxt);
            } else {
                _removeOne(nxt);
            }
            _rebalance(node);
        } else {
            _removeOne(node);
        }
    }

    private void _removeOne(TreeNode node) {
        TreeNode replacement = node.left != null ? node.left : node.right;
        if (node.parent != null) {
            if (isLeft(node)) {
                node.parent.left = replacement;
            } else {
                node.parent.right = replacement;
            }
            if (replacement != null) {
                replacement.parent = node.parent;
            }
            node.parent = null;
        } else {
            tree = replacement;
            if (replacement != null) {
                replacement.parent = null;
            }
        }
        node.left = null;
        node.right = null;
        node.parent = null;
        _rebalance(replacement);
    }

    private void _removeLeaf(TreeNode node) {
        if (node.parent != null) {
            if (isLeft(node)) {
                node.parent.left = null;
            } else {
                node.parent.right = null;
            }
            _rebalance(node.parent);
        } else {
            tree = null;
        }
        node.parent = null;
        node.left = null;
        node.right = null;
    }

    private TreeNode _getNode(int k) {
        if (tree == null) return null;
        TreeNode node = tree;
        while (node != null) {
            if (k < node.key) {
                node = node.left;
            } else if (node.key < k) {
                node = node.right;
            } else {
                return node;
            }
        }
        return null;
    }

    private static boolean isLeaf(TreeNode node) {
        return node.left == null && node.right == null;
    }

    private static TreeNode _getNext(TreeNode node) {
        if (node.right == null) return node.parent;
        TreeNode n = node.right;
        while (n.left != null) {
            n = n.left;
        }
        return n;
    }
}

class SegmentTree {
    private final int[] data;
    private final int size;
    private final int defaultValue;
    private final IntBinaryOperator func;

    public SegmentTree(int[] data, int defaultValue, IntBinaryOperator func) {
        this.defaultValue = defaultValue;
        this.func = func;
        this.size = 1 << (Integer.highestOneBit(data.length) << 1);
        this.data = new int[size];
        System.arraycopy(data, 0, this.data, size / 2, data.length);
        for (int i = size / 2 - 1; i > 0; i--) {
            this.data[i] = func.applyAsInt(this.data[i * 2], this.data[i * 2 + 1]);
        }
    }

    public void set(int idx, int value) {
        idx += size / 2;
        data[idx] = value;
        while (idx > 1) {
            idx /= 2;
            data[idx] = func.applyAsInt(data[idx * 2], data[idx * 2 + 1]);
        }
    }

    public int query(int start, int stop) {
        start += size / 2;
        stop += size / 2 + 1;
        int res = defaultValue;
        while (start < stop) {
            if ((start & 1) == 1) {
                res = func.applyAsInt(res, data[start++]);
            }
            if ((stop & 1) == 1) {
                res = func.applyAsInt(res, data[--stop]);
            }
            start /= 2;
            stop /= 2;
        }
        return res;
    }
}

class Combination {
    private final int MOD;
    private final Factorial factorial;

    public Combination(int MOD) {
        this.MOD = MOD;
        this.factorial = new Factorial(MOD);
    }

    public int ncr(int n, int k) {
        if (k < 0 || n < k) return 0;
        k = Math.min(k, n - k);
        return (int) ((long) factorial.calc(n) * factorial.invFactorial(Math.max(n - k, k)) % MOD * factorial.invFactorial(Math.min(k, n - k)) % MOD);
    }
}

class Factorial {
    private final int MOD;
    private final List<Integer> factorials = new ArrayList<>(Arrays.asList(1, 1));
    private final List<Integer> invModulos = new ArrayList<>(Arrays.asList(0, 1));
    private final List<Integer> invFactorial_ = new ArrayList<>(Arrays.asList(1, 1));

    public Factorial(int MOD) {
        this.MOD = MOD;
    }

    public int calc(int n) {
        if (n < 0) throw new IllegalArgumentException("n must be non-negative");
        if (n < factorials.size()) return factorials.get(n);
        int initialI = factorials.size();
        int prev = factorials.get(initialI - 1);
        for (int i = initialI; i <= n; i++) {
            prev = (int) ((long) prev * i % MOD);
            factorials.add(prev);
        }
        return factorials.get(n);
    }

    public int inv(int n) {
        if (n < 0) throw new IllegalArgumentException("n must be non-negative");
        if (n < invModulos.size()) return invModulos.get(n);
        int initialI = invModulos.size();
        for (int i = initialI; i < Math.min(MOD, n + 1); i++) {
            invModulos.add((int) (-invModulos.get(MOD % i) * (MOD / i) % MOD + MOD) % MOD);
        }
        return invModulos.get(n % MOD);
    }

    public int invFactorial(int n) {
        if (n < 0) throw new IllegalArgumentException("n must be non-negative");
        if (n < invFactorial_.size()) return invFactorial_.get(n);
        inv(n);
        int initialI = invFactorial_.size();
        int prev = invFactorial_.get(initialI - 1);
        for (int i = initialI; i <= n; i++) {
            prev = (int) ((long) prev * invModulos.get(i % MOD) % MOD);
            invFactorial_.add(prev);
        }
        return invFactorial_.get(n);
    }
}

public class 5cb6c1b3fc0e089985a249812459323b_nc {
    public static void main(String[] args) throws IOException {
        FastIO io = new FastIO(System.in, System.out);
        int n = Integer.parseInt(io.next());
        int k = Integer.parseInt(io.next());
        int[] l = new int[n];
        for (int i = 0; i < n; i++) {
            l[i] = Integer.parseInt(io.next()) % k;
        }
        if (n == 1) {
            io.println(l[0]);
            io.close();
            return;
        }
        int[] l1 = Arrays.copyOfRange(l, 0, n / 2);
        int[] l2 = Arrays.copyOfRange(l, n / 2, n);
        Set<Integer> s = new HashSet<>();
        Set<Integer> s1 = new HashSet<>();
        Set<Integer> s2 = new HashSet<>();
        for (int i : l1) {
            for (int j : s) {
                s2.add((j + i) % k);
            }
            s.add(i);
            s.addAll(s2);
            s2.clear();
        }
        for (int i : l2) {
            for (int j : s1) {
                s2.add((j + i) % k);
            }
            s1.add(i);
            s1.addAll(s2);
            s2.clear();
        }
        List<Integer> sList = new ArrayList<>(s);
        List<Integer> s1List = new ArrayList<>(s1);
        Collections.sort(sList);
        Collections.sort(s1List);
        int ans = Math.max(Collections.max(sList), Collections.max(s1List));
        for (int i = 0; i < sList.size(); i++) {
            int t = k - 1 - sList.get(i);
            int ind = binarySearchCount(s1List, t);
            int r = s1List.get(ind - 1);
            int rt = s1List.get(s1List.size() - 1);
            ans = Math.max(Math.max((r + sList.get(i)) % k, (sList.get(i) + rt) % k), ans);
        }
        io.println(ans);
        io.close();
    }

    private static int binarySearchCount(List<Integer> arr, int key) {
        int left = 0, right = arr.size() - 1, count = 0;
        while (left <= right) {
            int mid = (right + left) / 2;
            if (arr.get(mid) <= key) {
                count = mid + 1;
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        return count;
    }
}
